stan/math/rev/fun/dot_product.hpp

```diff
@@ -38,22 +39,27 @@ inline return_type_t<T1, T2> dot_product(const T1& v1, const T2& v2) {
   const auto& v1_col = as_column_vector_or_scalar(v1);
   const auto& v2_col = as_column_vector_or_scalar(v2);
 
-  const auto& v1_val = to_arena_if<!is_constant<T2>::value>(value_of(v1_col));
-  const auto& v2_val = to_arena_if<!is_constant<T1>::value>(value_of(v2_col));
+  const auto& v1_val_arena
+      = to_arena_if<!is_constant<T2>::value>(value_of(v1_col));
+  const auto& v2_val_arena
+      = to_arena_if<!is_constant<T1>::value>(value_of(v2_col));
 
-  var res(new vari(dot_product(v1_val, v2_val)));
+  var res(new vari(dot_product(
+      static_select<is_constant<T2>::value>(value_of(v1_col), v1_val_arena),
```

Comment:
I think we just write out all the if statements here. It's annoying duplicating stuff, but it's easier to read code:

```cpp
double res_val = 0.0;
if(is_constant<T1>::value) {
  res_val = dot_product(v1_val_arena, value_of(v2_col));
} else if(is_constant<T2>::value) {
  res_val = dot_product(value_of(v1_col), v2_val_arena);
} else {
  res_val = dot_product(v1_val_arena, v2_val_arena);
}
var res(res_val);
```